compose: Auto-restore last draft for recipient (if it exists).

evykassirer opened the pull request:
## Feedback from CZO

Reported bugs that need fixing:
* [x] "If you hit Enter to reopen the compose box just after sending a message, sometimes you get the message you just sent popped back into your compose box. I think this is a race with the server confirming the reply via local echo, so likely very easy to reproduce if one sets one's browser to "slow 3G" network or adds a `time.sleep` in the send-message code path on the server. We likely need to add a field on draft objects to track that they're being sent to avoid this failure mode." [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/auto-restore.20last.20draft/near/1758342)
* [x] If you're in a topic that you leaked a draft to (E.g., I had a message I partially wrote 3 days ago but abandoned), and you use Enter or clicking on a message to open the compose box to reply to something current in the topic and immediately start typing, it feels very jarring to have what you're typing be just appended onto the end of the old draft, without any space between the old and new content. [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/auto-restore.20last.20draft/near/1758344)
  * solution: Add a space after the draft in the composebox, in these cases
* [x] If you have some text that you're drafting, but while working on that text, see something in that topic that you need to reply to with a rapid-fire series of short messages, there isn't a great workflow available... because every time you send  one of those messages, your long half-finished draft pops into the compose box. [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/auto-restore.20last.20draft/near/1758344)
  * solution: once you've selected "Save draft and start a new message", your drafts for this conversation will not get auto-restored until you leave the current view.
* [x] only have drafts saved after this change be eligible for auto-population into the compose box, to make the transition period not feel like that

Potential updates that seem not blocking to me:

* "Do we need an explicit 'discard' icon ('trash'?) for avoiding this situation, assuming compositions are auto-saved?" [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/auto-restore.20last.20draft/near/1757050)
* "If you happen to have a draft matching the current conversation, and try to use `Enter` to open the compose box (planning to tab to the topic input to compose to a new topic, say), then you end up changing the recipient for your existing draft rather than getting an empty compose box for your new topic. In theory, using the `c` keyboard shortcut (or "New topic" button) would avoid this, but I did notice myself encountering this a couple times in the last day, and I do use the `c` shortcut sometimes, so ... anyone with that habit may run into that problem." [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/auto-restore.20last.20draft/near/1758347)
  * I personally think this one is fine to let people reprogram habits


---

## PR details

Fixes #11218, #17396, and #18555, building off of @Riken-Shah's work.

![Kapture 2023-07-18 at 16 56 35](https://github.com/zulip/zulip/assets/5634097/29ed6094-3015-4788-848f-607ba0ca6fb5)

functionality notes:

* loads most recent draft by timestamp
* will show the "start a new message" note if there's any existing draft for the recipient, including anything written in the composebox

updated screenshot with what it looks like now to start a new message:

<img width="274" alt="image" src="https://github.com/zulip/zulip/assets/5634097/ea7c6e14-bbf6-4b2a-bc4a-6ed09e962b00">

alya commented:
Thanks -- super helpful to be able to play with this!

One thing I noticed is that this restores the last created draft for the conversation, not the last edited draft. I think last edited (as per #18555) would be better, and would be consistent with the order of drafts in the Drafts menu.

timabbott commented:
I suspect the test failure is actually caused by this change -- I haven't seen it before and I think what it's seeing is the compose box not being empty when reopened in some code path.

timabbott commented on static/js/drafts.js:
```diff
@@ -101,6 +101,30 @@ export const draft_model = (function () {
+            .sort((draft_a, draft_b) => drafts[draft_a].updatedAt - drafts[draft_b].updatedAt);
+    };
+
+    exports.getDraftsIdByRecipients = function (private_message_recipient) {
```
Hmm, this class uses camelCase naming, so I guess adding this isn't a regression.

timabbott commented on static/js/drafts.js:
```diff
@@ -101,6 +101,30 @@ export const draft_model = (function () {
+                    drafts[draft_id].private_message_recipient.split(",").sort(),
+                    private_message_recipient.split(",").sort(),
+                ),
+        );
```
Why does this one not sort like the previous one does?

evykassirer replied:
I didn't write this commit, but it seems like it should also sort. Thanks for calling that out, I'll add it in!

evykassirer replied:
my bad for not going over the old commits very thoroughly before putting it up for review

timabbott commented on frontend_tests/puppeteer_tests/drafts.ts:
```diff
@@ -40,7 +40,7 @@ async function create_stream_message_draft(page: Page): Promise<void> {
     await page.waitForSelector("#stream-message", {visible: true});
     await common.fill_form(page, "form#send_message_form", {
-        stream_message_recipient_stream: "all",
+        stream_message_recipient_stream: "Verona",
```
This puppeteer test commit fails in isolation. It'd be best to reorder this to be a prep commit that passes in isolation, and then having whatever commit is changing something that this test relies on change the test to verify the new behavior.

```
AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:
+ actual - expected

+ ''
- 'Updated stream message'
    at test_edited_draft_message (/home/tabbott/zulip/frontend_tests/puppeteer_tests/drafts.ts:163:12)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async drafts_test (/home/tabbott/zulip/frontend_tests/puppeteer_tests/drafts.ts:257:5)
    at async CommonUtils.run_test_async (/home/tabbott/zulip/frontend_tests/puppeteer_lib/common.ts:619:13) {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: '',
  expected: 'Updated stream message',
  operator: 'strictEqual'
}

```

timabbott commented:
I merged d04385dd1f138f53f1f15c7ede8934cbc0cc4e0a; thanks @evykassirer.

As a next step, let's try to get the other 3-4 prep commits that conceptually should be correct in isolation integrated; posted comments on the main ones. Wouldn't be unreasonable to pull them out to a separate small PR so CI can check them for us.

evykassirer commented on frontend_tests/puppeteer_tests/drafts.ts:
```diff
@@ -40,13 +40,30 @@ async function create_stream_message_draft(page: Page): Promise<void> {
     await page.waitForSelector("#stream-message", {visible: true});
     await common.fill_form(page, "form#send_message_form", {
-        stream_message_recipient_stream: "all",
+        stream_message_recipient_stream: "Verona",
```
This needed to be updated to "Verona" to be a proper stream view to support the new tests that were added that involve the stream view and not just the drafts view.

evykassirer commented on frontend_tests/puppeteer_tests/drafts.ts:
```diff
@@ -103,7 +132,7 @@ async function test_previously_created_drafts_rendered(page: Page): Promise<void
             ".draft-row .message_header_private_message .stream_label",
         ),
-        "You and Cordelia, Lear's daughter, King Hamlet",
+        "You and King Hamlet, Cordelia, Lear's daughter",
```
I'm not totally sure why this was switched around, but my guess was that by actually sending a message to the stream instead of having this recipient only in the draft, the order got sorted/swapped?

evykassirer commented on frontend_tests/puppeteer_tests/drafts.ts:
```diff
@@ -179,7 +209,7 @@ async function test_restore_private_message_draft_via_draft_overlay(page: Page):
     const cordelia_internal_email = await common.get_internal_email_from_name(page, "cordelia");
     const hamlet_internal_email = await common.get_internal_email_from_name(page, "hamlet");
-    await common.pm_recipient.expect(page, `${cordelia_internal_email},${hamlet_internal_email}`);
+    await common.pm_recipient.expect(page, `${hamlet_internal_email},${cordelia_internal_email}`);
```
(same not-totally-sure situation as above)

amanagr commented on static/templates/draft_table_body.hbs:
```diff
@@ -22,6 +22,9 @@
                     {{#each narrow_drafts}}
                         {{> draft}}
                     {{/each}}
+                    <button type="button" id="compose-new-message-button" class="button sea-green">
```
I think we should keep `compose-new-message-button` as class and use `#draft_overlay .compose-new-message-button` to reference it so that I know where this button is placed regardless of the file I am reading.

evykassirer commented:
thanks!

amanagr commented on static/js/compose.js:
```diff
@@ -192,7 +192,7 @@ export function clear_compose_box() {
     if (compose_ui.is_full_size()) {
         compose_ui.make_compose_box_original_size();
     }
-    $("#compose-textarea").val("").trigger("focus");
+    $("#compose-textarea").trigger("focus").val("");
     compose_validate.check_overflow_text();
```
We should also add a comment for this change in the code.

evykassirer replied:
oh this actually isn't needed anymore! thanks

amanagr reviewed (commented):
The changes look good to me. Left 2 minor comments.

@alya @evykassirer  what are you thoughts on showing a tooltip when we do an `auto-restore` for the narrow?

<img width="611" alt="image" src="https://user-images.githubusercontent.com/25124304/201184374-745b7e0d-414f-4872-9bfc-4a74d9ac3395.png">

(maybe with a shorted text)

Also, we should have better contrast for the new compose message button so that it is the first thing user notices.

Something like this maybe:
<img width="674" alt="image" src="https://user-images.githubusercontent.com/25124304/201184554-ebef9251-91a6-42dd-9cfd-2fc6c362e60f.png">

evykassirer commented:
Tooltip sounds like a good idea to me! I'll wait on Alya's thoughts before implementing it

amanagr commented:
@alya FYI

alya commented:
Sorry for the delay! I'm a few days behind on reviews, and only saw myself tagged on Evy's checkin yesterday.

1. Sure, we can try a tooltip. Maybe something like, "Click to save message as a draft and start a new one." ? I think it's still quite subtle/hard to notice; I wonder if we should somehow highlight the "Drafts" link, e.g. by putting a blue dot next to it or something. (Regardless, this wouldn't be a blocker for test deployment.)

2. For the button in Drafts:
   - Let's put it at the top, even above the heading for the narrow. That should make it easier to find.
   - The current black bar design isn't working for me -- it doesn't match anything else in the app, and I think it's not even obvious that it's a button. Can we try our standard bluish-purple action button instead?

alya commented:
Functionality-wise, I noticed that the button is there when I go to drafts with an empty compose box. This seems like a bug, since the draft doesn't actually get saved if I click the button in that situation. We should only show it if the compose box has the minimum number of characters required for saving a draft.

alya commented:
Excited to see progress on this!

evykassirer commented:
how about "Click to save message and start a new one"? I think "as a draft" is implied fairly well

evykassirer commented:
Addressed all the comments except the tooltip. I think test deploying without it and seeing what people think feels reasonable. Thoughts?

amanagr commented:
> Click to save message and start a new one

"Click to possibly save message and start a new one" ?

So, that someone who just wants to open drafts doesn't get confused.

alya commented:
For the button, can we try making it the width of the text, rather than going all the way across? I think that will be more consistent with other buttons we have. I'm not sure whether centered or left-aligned would work best; the "Create stream" button (which is centered) might be the closet analogue.

<img width="575" alt="Screen Shot 2022-11-28 at 1 15 07 AM" src="https://user-images.githubusercontent.com/2090066/204239165-32a9204b-41e8-439b-8b94-4fced7d04213.png">

evykassirer commented:
I really feel like the tooltip will be confusing with either phrasing. Could we try test deploying without it, and see if people end up wanting to use the button (and seeing where they thought to look for it / what kind of hint they were expecting)?

evykassirer commented:
Made the other updates :)

alya commented:
> I really feel like the tooltip will be confusing with either phrasing. Could we try test deploying without it

Sure, that's fine by me.

alya commented:
> Made the other updates :)

Great! Could you please post an updated screenshot?

evykassirer commented:
screenshots updated!

alya commented:
Cool, thanks! Looks like a reasonable UI to try out to me.

@timabbott It would be great to get your review for CZO deployment.

evykassirer commented:
@timabbott anything I can do to help with review on this?

evykassirer commented:
last update is from Alya:

> Looks like a reasonable UI to try out to me.
>
> Tim, It would be great to get your review for CZO deployment.

next step is CZO deployment from Tim

alya commented:
@timabbott Would be great to test this out, so that we can decide whether it's realistic to include in the 7.0 release.

evykassirer commented:
thanks Alya! I'd really appreciate that

evykassirer commented:
I've changed where the "start new message" functionality goes, adding it to the new "schedule later" menu. Pushed that change and added fresh screencapture in the PR description

shameondev commented:
Hey @evykassirer, do you need any help with anything here?

karlstolley commented:
This is looking and feeling pretty good to me!

One edge case that I thought of, with two different parts to it (this is derived from how I tend to use drafts for writing my check-ins):

1. If someone has a draft written, but wants to send something else first, going to the triple-dots menu to find the `Start new message` button/link is a big ask, and non-obvious. I could easily imagine losing stuff that way. I'm not sure what the solution is here, but it seems like it should be more work to intentionally delete a draft, rather than it being more work to intentionally save a draft and start a new message.
2. But at the same time, if I have multiple drafts written on the same thread, it would be really cool if there were logic to see if any other drafts existed, and perhaps loaded them up? I'm not entirely sure how that would work, or how it would feel to hit Send and then discover that the message box has other, different content in it. But given that you can essentially have that by clicking away and then clicking the compose box open again, it does seem like there should be some kind of solution here?

evykassirer commented:
> how it would feel to hit Send and then discover that the message box has other, different content in it.

I'm pretty sure the functionality right now is that once you hit send, the compose box clears and doesn't load another draft. This makes sense to me because people sometimes send multiple messages in a row and I don't want to bring up old drafts in the middle of that.

> given that you can essentially have that by clicking away and then clicking the compose box open again, it does seem like there should be some kind of solution here?

Oh yeah, so that's how it works right now. Are you saying you'd prefer to have it load as soon as you send the message?

> it seems like it should be more work to intentionally delete a draft, rather than it being more work to intentionally save a draft and start a new message.

I think this will depend a lot on how people end up having drafts saved. Sometimes people will accidentally make drafts a lot, and sometimes people deliberately want to have several drafts written for a topic.

It also seems useful to keep in mind that people coming from Slack or Discord (or most other chat apps) have the default functionality expectation of only getting one draft per recipient and deleting it by removing the text. I think that Zulip's option to have multiple drafts is better, but part of why I chose the functionality the way I did was so that it could align with the way this works with other chat apps.

I also think it would be great to just deploy this to CZO and see what people think instead of hypothesizing too much about what users will prefer.

evykassirer commented:
> Hey @evykassirer, do you need any help with anything here?

Any code review or product input would be helpful! I'm trying to move this PR towards being tested on CZO.

karlstolley commented:
> I also think it would be great to just deploy this to CZO and see what people think instead of hypothesizing too much about what users will prefer.

I agree :) I don't have strong preferences for any of those things I mentioned. Just raising the case...but you're right that we just need to see how this would feel with some regular use on CZO.

evykassirer commented:
Yeah I appreciate your input! I look forward to seeing what folks on CZO think :)

evykassirer commented:
Anything else to do here, or do you think it's ready for integration review?

karlstolley commented:
@timabbott, do you have appetite for running this PR as an experiment on CZO? I know we've got a number of experiments running now, but I've labeled this as such for when it's convenient.

evykassirer commented:
@timabbott @alya this is rebased and ready for checking on on CZO!

timabbott commented on web/src/drafts.js:
```diff
@@ -274,6 +275,62 @@ export function update_draft(opts = {}) {
+                drafts[draft_id].topic === topic,
+        )
+        .sort((draft_a, draft_b) => drafts[draft_a].updatedAt - drafts[draft_b].updatedAt);
+}
```
This is a case-sensitive topic comparison; should this be case-insensitive, perhaps using `util.same_topic`?

timabbott commented on web/src/drafts.js:
```diff
@@ -274,6 +275,62 @@ export function update_draft(opts = {}) {
+                _.isEqual(
+                    drafts[draft_id].private_message_recipient.split(",").sort(),
+                    private_message_recipient.split(",").sort(),
+                ),
```
Is the splitting/sorting the thing to do here? I would have thought that the `private_message_recipient` fields were already a sorted, comma-separated list of `user_id` values, and so comparing as strings should have the exact same result.

timabbott commented on web/src/drafts.js:
```diff
@@ -274,6 +275,62 @@ export function update_draft(opts = {}) {
+    const message_type = compose_state.get_message_type();
+    if (!message_type || compose_state.message_content()) {
+        return;
+    }
```
Let's make these two separate if statements, with comments noting when they are expected to be the case. 

Actually the `message_type` check might deserve to be an `assert`? `compose_state.set_message_type(msg_type);` is called just before this function.

timabbott commented on web/src/compose_actions.js:
```diff
@@ -270,6 +270,12 @@ export function start(msg_type, opts) {
 
     compose_state.set_message_type(msg_type);
```
I think it might be fruitful to move this call above the `opts.content` check, and have your check for this feature sit as an else clause to `opts.content !== undefined`, which is the code path used by "quote and reply" and similar features, which I expect should always .

Hmm, `compose_reply.quote_and_reply` doesn't use that feature.

evykassirer replied:
@timabbott FYI waiting on your response to this to move forward

timabbott replied:
This feels very messy. I think it might be better to just have an option passed in the small number of code paths where the intent is to start with a fresh messy, rather than adding an option to all of the places we open compose?

evykassirer replied:
> where the intent is to start with a fresh messy

Is this supposed to say fresh draft? As in, only pass the option when we _don't_ want to load a previous draft?

Yeah I agree it's pretty messy, which I don't like. The reason I ended up doing it this way is that it's not always obvious that drafts are being overwritten, so future situations where we add code to open the compose box and then overwrite it, it'll be really important to make sure we're not loading a previous draft. If loading a previous draft happens by default, then it just makes it easier for that to happen.

But I'm happy to switch it to only pass the option sometimes, if that seems like the better tradeoff. I just want to be clear about the tradeoff.

timabbott replied:
Yeah, "fresh draft". Correct, we'd pass the option when we don't want to load a draft. I think there's be at most a handful of places that needs to be passed.

I'm not sure I understand the point about drafts being overwritten -- if you see a draft in the compose box and replace it with other text, I think you'd expect that the old text is wrong. E.g. gmail drafts work the same way.

evykassirer replied:
The overwriting doesn't happen from user choice, but from code.

If we forgot to pass this option for quote-replying, and there was a draft we could auto-load, then that draft would automatically be overwritten when we replaced the compose content with the quote reply, and the draft would be lost. So it's very important to remember to pass the option for starting a fresh draft when it's relevant, and if we forget then drafts can get overwritten (essentially deleted).

timabbott commented on web/src/compose_actions.js:
```diff
@@ -270,6 +270,12 @@ export function start(msg_type, opts) {
 
     compose_state.set_message_type(msg_type);
 
+    if (opts.trigger !== "restore from drafts menu") {
```
Is this condition equivalent to `opts.draft_id` being passed?

timabbott commented on web/src/drafts.js:
```diff
@@ -274,6 +275,62 @@ export function update_draft(opts = {}) {
+
+    compose_state.message_content(draft.content);
+    compose_ui.autosize_textarea($("#compose-textarea"));
+    $("#compose-textarea").trigger("select").data("draft-id", last_draft_id);
```
What do you think about having this function just return a `draft_id` or undefined, and the compose code can then add that `draft_id` to `opts` and process accordingly?

That seems like it'd be much more obviously producing an equivalent result to having restored the draft another way, and would let us remove these violations of the `compose_actions` abstraction boundary that it's in charge of adjusting the compose UI.

evykassirer replied:
That sounds like a great idea, thanks. I hadn't considered that quote reply and whatnot would overwrite restored drafts, so I'm really glad you thought of that.

timabbott commented on web/src/drafts.js:
```diff
@@ -274,6 +275,62 @@ export function update_draft(opts = {}) {
+        if (stream_id !== undefined && topic !== "") {
+            last_draft_id = get_draft_ids_for_topic(stream_id, topic).pop();
+        }
+    }
```
What's the motivation for having these two functions, rather than just a `get_last_drop_id_for_compose_state`?

Actually never mind, we have `filter_drafts_by_compose_box_and_recipient`, which is used in the drafts overlay and does this exact same calculation, but without the bugs noted above. So I propose just moving that function to `drafts.js`, and calling it to find the drafts to consider.

evykassirer replied:
Great catch, thanks. Switching to that

timabbott commented:
Sorry for the delay in reviewing this. About half the comments would be addressed by https://github.com/zulip/zulip/pull/22104#discussion_r1481984162 which suggests reusing existing correct logic, so I recommend starting with that comment.

evykassirer commented:
I addressed the easy comments, but there's one big issue I'm not quite sure how to address. I think the thing I just pushed works fine, but it might also be too brittle.

timabbott commented on web/src/drafts.js:
```diff
@@ -274,6 +275,78 @@ export function update_draft(opts = {}) {
+    }
+    return _.pick(drafts, narrow_drafts_ids);
+}
+
```
Opened #28923 to merge this move, since that'll reduce merge conflicts with other work.

evykassirer replied:
thanks!

evykassirer commented:
@timabbott this is ready for you to look at again!

timabbott commented on web/src/scheduled_messages_popover.js:
```diff
@@ -150,11 +152,21 @@ export function initialize() {
+            const show_compose_new_message =
+                narrow_drafts &&
+                compose_state.has_message_content() &&
+                compose_state.has_complete_recipient();
```
I don't understand the conditional here -- shouldn't this option be useful regardless of whether you have any drafts matching the view? Like what if you just have some text in the compose box right now (not already saved as a draft) and want to start afresh saving that content as a draft?

evykassirer replied:
I'll change this when I make the other change, but I'll wait for your response on that first

evykassirer replied:
I changed to this

```js
            // If there's existing text in the composebox, show an option
            // to keep/save the current draft and start a new message.
            const show_compose_new_message = compose_state.has_message_content();
```

evykassirer replied:
I actually ended up making a `has_savable_message_content`, which checks if the draft is long enough to save, because I don't want to make false promises of saving a draft when we actually throw away "draft" messages of <= 2 characters.

timabbott commented:
Replied to the review questions.

evykassirer commented on web/src/compose_state.ts:
```diff
@@ -194,6 +194,13 @@ export function has_message_content(): boolean {
     return message_content() !== "";
 }
 
+export const MINIMUM_MESSAGE_LENGTH_TO_SAVE_DRAFT = 2;
```
I feel like ideally this would be in `drafts.js` but then there would be a circular import situation

evykassirer commented on web/src/compose_state.ts:
```diff
@@ -194,6 +194,13 @@ export function has_message_content(): boolean {
+export const MINIMUM_MESSAGE_LENGTH_TO_SAVE_DRAFT = 2;
+export function has_savable_message_content(): boolean {
+    return (
+        message_content() !== "" && message_content().length > MINIMUM_MESSAGE_LENGTH_TO_SAVE_DRAFT
```
Also open to a shorter name for it to avoid this wrapping

timabbott replied:
I think the `""` has a short length, so this doesn't need to wrap if you just delete that first redundant check.

evykassirer replied:
Oh of course 😅

evykassirer commented:
I made the remaining updates! Should be ready for review again

timabbott commented on web/src/drafts.ts:
```diff
@@ -474,6 +474,26 @@ export function filter_drafts_by_compose_box_and_recipient(
 }
 
+export function get_last_draft_based_on_compose_state():
+    | (LocalStorageDraft & {id: string})
```
Might be worth giving this type a nice exported name like `Draft`. I guess that type exists as `LocalStorageDraftWithId` today.

evykassirer replied:
Ah great, replaced with that

timabbott commented on web/src/drafts.ts:
```diff
@@ -474,6 +474,26 @@ export function filter_drafts_by_compose_box_and_recipient(
+    | undefined {
+    if (compose_state.message_content()) {
+        return undefined;
+    }
```
Why is this check here? It feels to me like this should be the caller's responsibility.

evykassirer replied:
I can't remember at this point, and it makes sense to me to make it the caller's responsibility so I'll just move it there

timabbott commented on web/src/compose_reply.js:
```diff
@@ -27,7 +27,10 @@ export function respond_to_message(opts) {
-            compose_actions.start("stream", {trigger: "recent_view_nofocus"});
+            compose_actions.start("stream", {
+                trigger: "recent_view_nofocus",
+                keep_composebox_empty: opts.keep_composebox_empty,
```
Just to confirm I understand correctly: This option is a noop in this case, right, because there won't be specified stream/topic, and thus `drafts.get_last_draft_based_on_compose_state` should return nothing? 

I think it's probably correct to have the option anyway.

evykassirer replied:
Yeah, that sounds right to me. To be clear, I'm defaulting in this function to just always pass through `opts.keep_composebox_empty` to `compose_actions.start` and didn't think about each case super throroughly, since I trust the caller to this function to make the decision. I'm open to being more stingy/cautious about it if you prefer that though.

timabbott replied:
Yeah, I think the thing I'd wonder about is whether we should be passing through some subset of `opts` more generically, but I'm not sure I want to spend time on fiddling with that.

timabbott commented on web/src/compose_reply.js:
```diff
@@ -64,7 +71,10 @@ export function respond_to_message(opts) {
+                    keep_composebox_empty: opts.keep_composebox_empty,
+                });
                 return;
             }
```
I don't understand this case at all - the existing logic seems crazy. I don't see why we're looking at the first term rather than just the `stream` operand, if any. But maybe it's a case we'll delete in https://github.com/zulip/zulip/pull/21852?

timabbott commented on web/src/scheduled_messages_ui.js:
```diff
@@ -59,6 +59,7 @@ export function open_scheduled_message_in_compose(scheduled_msg, should_narrow_t
             type: scheduled_msg.type,
             private_message_recipient: recipient_emails.join(","),
             content: scheduled_msg.content,
+            keep_composebox_empty: true,
```
This option seems like it's not required; the `content` option is doing the job for us, right?

evykassirer replied:
`content` should already handle this case yes, but I was imagining it might be clearer to also pass `keep_composebox_empty` more explicitly? And also maybe would help anyone doing similar things in the future remember about making sure the composebox is kept empty. 

I don't feel super strongly about having or not having it there. Let me know if you prefer I remove it.

timabbott replied:
Yeah, it's fine for it to be there, just wanted to make sure I understood how this works correctly.

timabbott commented on web/src/compose_reply.js:
```diff
@@ -118,6 +131,7 @@ export function respond_to_message(opts) {
+        keep_composebox_empty: opts.keep_composebox_empty,
     });
 }
 
```
I think the quote-and-reply function is missing this?

``` diff
diff --git a/web/src/compose_reply.js b/web/src/compose_reply.js
index f8e68ccef1..da1d0d9c51 100644
--- a/web/src/compose_reply.js
+++ b/web/src/compose_reply.js
@@ -163,7 +163,7 @@ export function quote_and_reply(opts) {
         // are prone to glitches where you select the
         // text, plus it's a complicated codepath that
         // can have other unintended consequences.)
-        respond_to_message(opts);
+        respond_to_message({...opts, keep_composebox_empty: true});
     }
 
     compose_ui.insert_syntax_and_focus(quoting_placeholder, $textarea, "block");
```

evykassirer replied:
Oh shoot yes, that one's important! Thanks.

It was also missing from `reply_with_mention` so I added it to that as well.

I think I was grepping for `.respond_to_message` so I missed the calls within the module `compose_reply.js` but that should be all fixed now.

timabbott replied:
Makes sense. I really wish there was a way to make `git grep` find those local calls too, but you really need something fancier for it.

evykassirer reviewed (commented):
Thanks for the review! I've addressed (or at least replied to) all comments :)

timabbott commented on web/src/compose.js:
```diff
@@ -246,6 +250,10 @@ export function send_message(request = create_message_object()) {
+
+        const draft = drafts.draft_model.getDraft(request.draft_id);
+        draft.is_sending_saving = false;
+        drafts.draft_model.editDraft(request.draft_id, draft);
```
So I think we probably need some added change that `drafts.initialize` will st `is_sending_saving` to `false` for every draft as it loads them, at least if the draft isn't in the process of being sent?

evykassirer replied:
It seems like if `send_after_reload` is `true`, then we call `compose.finish`, which will set `is_sending_saving` to `true` again and then eventually set it to `false` once it completes. This doesn't happen for every reload, but specifically for a call to `reload.initiate({` when we think we need to reload the server in the middle of a message send that's timing out.

I see the concern of having drafts indefinitely stuck in sending limbo if they get interrupted by a reload, and that can definitely happen if the user just refreshes manually in the middle of a message sending. Thanks for flagging that!

I'm not sure how we'd know which drafts are "in the process of being sent by some on-startup logic" in `drafts.initialize`. I was wondering if maybe we could reset in `drafts.initialize` and then have the reload code run after that, but it seems like `drafts.initialize` is called _after_ the reload setup happens. The only other idea I have right now is to set a separate flag for sending during reload, which feels a bit bulky and messy to me, but maybe that's the best option. I'm curious if you have any ideas around how to do this?

timabbott replied:
I don't se a reason we couldn't do the data part of `drafts.initialize` quite early. To make that work, I think we'd just have to split `drafts.initialize` such that `set_count(Object.keys(draft_model.get()).length);` was in a separate `drafts.initialize_ui()` step that runs later.

A potentially good preparatory refactor we could merge the simplify this would be to actually split the module; a new `draft_counts_ui.ts` would have `set_counts`, and that function would be passed into `drafts.initialize` like the on-narrow hooks are to avoid import cycles. But I think the little split suggested here would suffice to resolve this issue.

evykassirer replied:
Ah great, thanks. I made an `initialize_counts` function to split this out.

alya commented:
@timabbott I think this is ready for an updated deployment.

evykassirer commented:
Rebased. The state of this PR is waiting for review on the last few commits, and also waiting on a response from Tim on the above thread.

timabbott commented on web/src/narrow.js:
```diff
@@ -359,6 +359,9 @@ export function activate(raw_terms, opts) {
         reset_ui_state();
 
+        // We sometimes prevent draft restoring until the narrow resets.
+        compose_state.allow_draft_restoring();
```
This could probably reasonably go in `reset_ui_state`... As could `compose_banner.clear_message_sent_banners();`, now that look at it.

evykassirer replied:
Great point, moved it and added a commit for moving `clear_message_sent_banners`

timabbott commented on web/src/scheduled_messages_popover.js:
```diff
@@ -210,6 +210,14 @@ export function initialize() {
+                compose_state.prevent_draft_restoring();
                 compose.clear_compose_box();
                 instance.hide();
             });
```
This file probably deserves to be renamed to `compose_send_menu_popover`

timabbott commented:
The fixup commits all seem fine to me aside from the notes above. Some seem like they might be best as independent commits rather than being squashed, but I'm about to go AFK so can't make a detailed suggestion.

evykassirer commented:
All comments should be addressed now! I squashed one of the fixup commits but left most of them standalone.

timabbott commented on web/src/drafts.ts:
```diff
@@ -281,7 +281,10 @@ export function rename_stream_recipient(
-    if (!compose_state.composing() || compose_state.message_content().length <= 2) {
+    if (
+        !compose_state.composing() ||
+        compose_state.message_content().length <= compose_state.MINIMUM_MESSAGE_LENGTH_TO_SAVE_DRAFT
```
This could be `!composet_state.has_savable_message_content`, right?

evykassirer replied:
Ah yes, thanks!

timabbott commented on web/src/scheduled_messages_popover.js:
```diff
@@ -203,6 +208,11 @@ export function initialize() {
+            $popper.one("click", ".compose_new_message", () => {
+                drafts.update_draft();
+                compose.clear_compose_box();
+                instance.hide();
```
This should be `popover_menus.hide_current_popover_if_visible(instance);`

timabbott commented on web/templates/popovers/send_later_popover.hbs:
```diff
@@ -53,4 +53,10 @@
+    {{#if show_compose_new_message}}
+    <hr />
+    <li>
+        <a class="compose_new_message" tabindex=="0">{{t "Save draft and start a new message"}}</a>
```
This should be `e" }}`

timabbott replied:
Hmm, it's less uniform than I'd have thought:
```
$ git grep '{{t "' | grep '"}}' | wc
    104     719   11614
$ git grep '{{t "' | grep '" }}' | wc
    531    4138   68899
```

May be worth discussing in #frontend, but we could also decide we're likely to reformat these templates eventually.

evykassirer replied:
Yeah I generally prefer to only have formatting rules like this that are enforced with a linter.

timabbott commented:
OK I addressed the small comments above, squashed the one fixup commit, and merged this, thanks for all the work on this @evykassirer!
